fix: mark GameServer Unhealthy when the game container terminates and the Pod stays Running (#4663)
googleforgames/agones: site/content/en/docs/Guides/health-checking.md

1. The Pod is set to `restartPolicy: Never` by default.
2. The SDK server sidecar container is set to `restartPolicy: Always`.
3. If a main container within the Pod fails, the `GameServer` will move to an `Unhealthy` state.

> 4. Since the Pod is `restartPolicy: Never`, a terminated game server container can never be restarted, so this happens as soon as the game server container exits with a non-zero exit code, rather than waiting for the Pod to reach the `Failed` phase.

Reviewer: ```suggestion
4. Assuming the Pod is `restartPolicy: Never`, a terminated game server container can never be restarted, so this happens as soon as the game server container exits with a non-zero exit code.
```

Depends on what people set it as. Also the "rather" isn't needed, since we don't have to compare it to the previous version.

   * If the game server container exits cleanly (exit code `0`), it is treated as a normal shutdown rather than a failure, and the `GameServer` moves to `Shutdown` once the Pod completes.

>    * The SDK server sidecar container stays alive for the entire duration of the Pod, and therefore SDK functionality is always available.

Reviewer: Thought: Should these be sub items or are they better as a 5 and 6?

7. If the SDK sidecar fails, then it will be restarted, assuming the `restartPolicy` remains the default.

### Running Additional Workloads Alongside the Game Server

If you run supporting workloads in the same Pod as your game server, such as log shippers, metrics agents or crash

> dump uploaders, declare them as

Reviewer: Nit"

```suggestion
dump uploaders, we recommend declaring them as
```

Just switching language to be best practices, rather than prescribed.

[Kubernetes sidecar containers](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/): entries in
`initContainers` with `restartPolicy: Always`, rather than extra entries in `containers`.

Kubernetes ties the Pod lifecycle to its *main* containers: a Pod only reaches the `Failed` or `Succeeded` phase once
**every** container in `containers` has terminated, while sidecar containers are
[terminated automatically](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/#sidecar-containers-and-pod-lifecycle)
once the main containers have exited. Declaring supporting workloads as sidecars therefore keeps the Pod's shutdown
semantics tied to the lifetime of the game server container:

* the game server container exiting is enough to end the Pod, and
* each sidecar is sent `SIGTERM` *after* the game server container has exited, and has the remainder of
  `terminationGracePeriodSeconds` to finish any work still in flight.

If you declare these workloads in `containers` instead, a long-lived one will hold the Pod in the `Running` phase after

> the game server container has died. Agones will still move the `GameServer` to `Unhealthy` as described in rule 3

Reviewer: ```suggestion
if the game server container has crashed or exited unexpectantly. Agones will still move the `GameServer` to `Unhealthy` as described in rule 3
```

Correct me if I'm wrong, but this is primarily an issue in the above scenario? If someone calls `SDK.Shutdown()` then everything works correctly, yeah?
